Add Language Server Protocol for GDScript

Geequlim opened the pull request:
It basicly works but not finished yet.

Add a plugin for godot editor to provide  a [LSP](https://microsoft.github.io/language-server-protocol) Server for GDScript language.

More features would be added in the future by @ankitpriyarup  who's the selected student of this slot of the GSoC this year.

This PR contains a commit of #29626 for the code completion usage.

I submit the PR to master. Once it was meged @ankitpriyarup could add  other features by submit other PRs based on this to master branch.

UPDATE: 2019-06-30
------------
The LSP implementation is works correctly both the server and client side.

Current supported features:
- document symbols
- workspace symbols
- diagnostics
- completion
- completion resolve
- hover
- definition
- **Smart resolve mode**:  to resolve symbols for dynamic typed variables to seed up coding.
- Script API dump

Planning features:
- signatureHelp
- documentLink

Testers are wellcomed. Here is the client for Visual Studio Code extension
[godot-tools-0.9.0.zip](https://github.com/godotengine/godot/files/3329595/godot-tools-0.9.0.zip)


![ezgif-2-81c8438ce5be](https://user-images.githubusercontent.com/6964556/60088350-1ae7ac80-9771-11e9-856e-8aa0e9ad35f3.gif)

Geequlim committed:
Add JSONRPC-2.0 implementation as a module

Geequlim committed:
Add GDScript Language Protocol plugin

groud commented on core/script_language.h:
```diff
@@ -200,6 +200,34 @@ class ScriptInstance {
 };
 
+struct ScriptCodeCompletionOption {
+	enum Kind {
```
As for you other PR, "Kind" is not the correct word here but "Type".

Geequlim replied:
I'm not really good at English.
I just followed the LSP `CompletionItem` https://code.visualstudio.com/api/references/vscode-api#CompletionItem

Geequlim commented:
Note for reviewer:
This PR is ready to review but not ready to be merged before #29781.
It contains the commit of #29626 so it need to be rebased after #29626 got merged.

Jason0214 commented on modules/jsonrpc/jsonrpc.cpp:
```diff
@@ -0,0 +1,171 @@
+	BIND_ENUM_CONSTANT(InternalError)
+}
+
+Dictionary JSONRPC::make_response_error(int p_code, const String &p_message, const Variant &p_id) const {
```
I would make function can be rewrite to a class like `JSONRPC_Error_Response`, and are the dict keys are made as class member and has a member function to convert it a dict or directly serialize it.  Anyway just some subjective opinions

Jason0214 reviewed (commented):
Just did a simple review based on the code style stuff. I have tested this branch and functionally works well.

Geequlim commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -0,0 +1,172 @@
+	if (OK == peer->get_packet_buffer(data)) {
+		String message;
+		message.parse_utf8((const char *)data.read().ptr(), data.size());
+		if (message.begins_with("Content-Length:")) return;
```
Just ignore this message as we can handle the length properly.

Geequlim commented on modules/gdscript/language_server/lsp.hpp:
```diff
@@ -0,0 +1,1293 @@
+/*************************************************************************/
```
I don't want the `lsp::hpp` contains so much constructors I'd like remove the default value assignments.

Geequlim replied:
The `lsp.hpp` was written in godot code by myself. Some of the ducumentations are coped from`https://github.com/Microsoft/language-server-protocol/blob/gh-pages/specification.md`. So I don't want to add more thirdparty files and messup the build scripts.

Geequlim commented on modules/gdscript/language_server/gdscript_extend_parser.cpp:
```diff
@@ -0,0 +1,237 @@
+	const int line = p_func->line - 1;
+	r_symbol.range.start.line = line;
+	r_symbol.range.start.character = p_func->column;
+	r_symbol.range.end.line = MAX(p_func->body->end_line - 2, p_func->body->line);
```
As the parser always give me a larger result.

vnen replied:
Is this because of the space between the functions (as in: the end line is the beginning of the next function) or the parser just gives an off-by-two value? Either way, I would prefer to fix this in the GDScript parser if possible.

Geequlim commented on modules/gdscript/language_server/gdscript_extend_parser.cpp:
```diff
@@ -0,0 +1,237 @@
+		lsp::Position pos;
+		int line = get_error_line() - 1;
+		const String &line_text = get_lines()[line];
+		pos.line = line;
```
That sounds good.

Geequlim committed:
Add a symbol pool to cache all native symbols and workspackes symbols.
Implement hover
Implement completion documentation resolve
Implement hover documentation
Implement go to definition

Geequlim committed:
Add optional smart resolve sulotion
The smart resolvaion can guess most symbols but it might be slow so disabled by default users can turn on it in the editor setting

Geequlim committed:
Improved performance for completion and symbol resolvation.
Improved uri and workspace path translatation on windows platform.
The smart resolvation is much faster than builtin's in the server side.
The smart resolve mode is still disabled as default as the clients might be slow with a planty of completion items.

Geequlim committed:
Improve symbol resolve for inner classes
Only level one inner classes would be resolved currently but it sould cover most real world use case
Improve documation parseing for const values
Improve documation format for native symbols

Geequlim committed:
Check client workspace directory  is valid
Drop test initialize message sent to client
Remove unused code property for the parser class

Geequlim committed:
Add generate script api to dictionary support
Expose GDScriptLanguageProtocol singleton and classes for editor plugins (Not visiable in class tree)
Fix minor bug in symbol resolve

Geequlim committed:
Add optional goto definition support for native symbols
This action will show help for target symbol in godot editor and bring the godot editor window to foreground
Improved markdown documentation for symbols.

akien-mga commented:
Should be rebased now that #29626 is merged.

Geequlim commented:
@akien-mga  Removed the first commit and rebased to master branch.
I'm not sure the commits are really correct.

Soaku commented on modules/gdscript/language_server/gdscript_language_protocol.h:
```diff
@@ -0,0 +1,86 @@
+#include "modules/websocket/websocket_peer.h"
+#include "modules/websocket/websocket_server.h"
+
+class GDScriptLanguageProtocol : public JSONRPC {
```
I think the best name would be `GDScriptLanguageServer`, the `Implementation` isn't necessary, alternatively `GDScriptLSP`

henriquelalves commented:
I'm testing this PR and it's overall working pretty well, but I'm having a weird crash using auto-complete when writing a "_init" constructor. To reproduce, just start writing "_init" on a newly created script and use auto-complete; it will write an extra "(" at the end (not sure if it's relevant for the crash), than state a "Expected '(' after identifier (syntax: 'func <identifier>([arguments]):'" error on VSCode, and saving the file will crash the engine. Godot will return this error on console: https://gist.github.com/henriquelalves/594746c6df44540567552eb7b8294c09. 

Also, I'm not sure if it was supposed to be working, but contextual node auto-completion (e.g. "get_node([..]" is only working for singletones, and F2 (renaming a symbol) doesn't work (apparently it's a Language Server method that has yet to be implemented, but I wasn't sure).

I'm going to test this PR a lot and I would love to help as much as I can with feedback, but is here the most appropriate channel? The branch this PR comes from doesn't have the "Issues" tab enabled, and I'm unsure if those are errors already being taken in consideration.

Geequlim committed:
Improve code compeletion for virtual methods with signatures

Geequlim commented:
@henriquelalves Thanks for your testing.
The issue with virtual method completion is fixed now, you need update the code and recompile the engine.

> Also, I'm not sure if it was supposed to be working, but contextual node auto-completion (e.g. "get_node([..]" is only working for singletones

The LSP server doesn't resolve node pathes as I think I not really neccessary because we code in external text editor with LSP client and it is really often we open diffrent script files in the external editor which is not attached to the opening scene in the godot editor so the node pathes from would be wrong.
I suggest you copy node pathes from the godot editor scene tree dock with `Ctrl + C` with target node.

> F2 (renaming a symbol) doesn't work (apparently it's a Language Server method that has yet to be implemented, but I wasn't sure).

GDScript is a dynamic typed language so it is hard to track references for symbols. So I'm not going to implement this feature as it may cause bugs randomly.
I suggest you use search and replace to do this kind of operations manually to avoid errors.

Any more feedbacks are welcomed. You can pin me here or in IRC.

henriquelalves commented:
> The LSP server doesn't resolve node pathes as I think I not really neccessary because we code in external text editor with LSP client and it is really often we open diffrent script files in the external editor which is not attached to the opening scene in the godot editor so the node pathes from would be wrong.
I suggest you copy node pathes from the godot editor scene tree dock with Ctrl + C with target node.

I still believe that this would be a useful feature for really big scenes, so instead of going back and forth from the editor to Godot, you could just focus on the editor itself - but I understand it's not necessary to make the feature go live. 

> The issue with virtual method completion is fixed now, you need update the code and recompile the engine.

I just recompiled with the last commit and it's still happening tho! But after I autocomplete any virtual method, it now finishes with a "()" (e.g. "func _input(event: InputEvent) -> void:()") and the editor crashes in the same way as before.

I'll keep testing as much as I can, as I'm already having a lot of fun coding in VSCode again! I just missed my plugins so much.

Jason0214 commented:
> The LSP server doesn't resolve node pathes as I think I not really neccessary because we code in external text editor with LSP client and it is really often we open diffrent script files in the external editor which is not attached to the opening scene in the godot editor so the node pathes from would be wrong.

Probably we can make this feature dispatchable at runtime, for example, let the plugin parsing some .tscn file in the workspace, if it can build a tree(subtree) of the script file, we can make use of it, otherwise we just don't care about node path. It could be very useful.
@ankitpriyarup has a implementation for renaming, it's actually promising, but also encounter the issue of external reference through node path.

@henriquelalves thanks for the testing! It is glad to see people interested in this work😄 

---
@Geequlim maybe we should push it a little bit to get more reviews on this PR. I will also tell @ankitpriyarup to merge his work into this branch (he has a doc support, partly working renaming and some fixes to autocompletion).

vnen commented on modules/gdscript/language_server/gdscript_extend_parser.cpp:
```diff
@@ -0,0 +1,759 @@
+
+String ExtendGDScriptParser::get_text_for_completion(const lsp::Position &p_cursor) const {
+
+	String longthing;
```
The `StringBuilder` class might be more suitable for this kind of string concatenation.

vnen commented on modules/gdscript/language_server/gdscript_workspace.cpp:
```diff
@@ -0,0 +1,504 @@
+
+void GDScriptWorkspace::list_script_files(const String &p_root_dir, List<String> &r_files) {
+	Error err;
+	DirAccessRef dir = DirAccess::open(p_root_dir, &err);
```
Traversing the file system might be really slow on large project. I think it would be better to get the `EditorFileSystem` cache.

Geequlim replied:
This is running in a seperated thread so I think it would be ok for performance.

vnen commented:
Took a glance at the code and it looks mostly okay, just pointed out a couple of things that I commented.

Since this is mostly new code I believe it's okay to be merged, as it shouldn't affect any current feature. We can fix potential issues with LSP itself later on.

Schroedi commented:
Thanks for tackling this important task! When it works, it is fantastic. It sometimes seems to crash though and has some minor annoyances. As the bugs in the vscode client repository are disabled, I post them here as well.

### Symptoms/Problems
#### Cyclic references disable completions
Sometimes godot does not detect the cyclic reference or the message is only shown later. It is obvious when the error message is there, but this is not always the case. No completion will be shown afterward. In I guess any file that is affected? The message is only shown in one file for me in a larger project, but the completion seems to be broken everywhere. In the attached sample, the completion works in unaffected files.
[LspTest.zip](https://github.com/godotengine/godot/files/3550374/LspTest.zip)


#### Sometimes godot hangs while closing
I am unable to reproduce it reliably. It seems to happen when the lsp crashed. I cannot reproduce it in the minimal sample.

#### Syntax highlighting is for CamelCase variables and functions is strange
![Screenshot from 2019-08-28 11-05-03](https://user-images.githubusercontent.com/97124/63841841-d3091f80-c983-11e9-9243-325ad0044ddb.png)

#### Starting godot from vscode fails on Linux 
In godot-tools.ts:78 I had to replace windows by linux:
let shell = vscode.workspace.getConfiguration("terminal.integrated.shell").get("linux", "");
And explicitly set terminal.integrated.shell.linux to a value (bash and zsh both worked).
Otherwise, godot open workspace in editor and godot start project would fail.
Without the modification I get this error from code --verbose when clicking on the open editor message in the failed to connect popup.
```
[1868:0828/122706.909802:INFO:CONSOLE(240)] "%c[Extension Host] %crejected promise not handled within 1 second: TypeError: Cannot read property 'endsWith' of null %c(at t.onMessage.process.on.t.catch.t (/usr/lib/code/out/vs/workbench/services/extensions/node/extensionHostProcess.js:759:23))", source: file:///usr/lib/code/out/vs/workbench/workbench.main.js (240)
[1868:0828/122706.910095:INFO:CONSOLE(240)] "%c[Extension Host] %cstack trace: TypeError: Cannot read property 'endsWith' of null
	at escape_command (/home/ascadian/.vscode-oss/extensions/geequlim.godot-tools-0.9.0/out/godot-tools.js:60:31)
	at run_godot (/home/ascadian/.vscode-oss/extensions/geequlim.godot-tools-0.9.0/out/godot-tools.js:70:34)
	at run_editor.Promise (/home/ascadian/.vscode-oss/extensions/geequlim.godot-tools-0.9.0/out/godot-tools.js:96:17)
	at new Promise (<anonymous>)
	at GodotTools.run_editor (/home/ascadian/.vscode-oss/extensions/geequlim.godot-tools-0.9.0/out/godot-tools.js:55:16)
	at open_workspace_with_editor.Promise (/home/ascadian/.vscode-oss/extensions/geequlim.godot-tools-0.9.0/out/godot-tools.js:45:22)
	at new Promise (<anonymous>)
	at GodotTools.open_workspace_with_editor (/home/ascadian/.vscode-oss/extensions/geequlim.godot-tools-0.9.0/out/godot-tools.js:38:16)
	at retry_connect_client.vscode.window.showErrorMessage.then.item (/home/ascadian/.vscode-oss/extensions/geequlim.godot-tools-0.9.0/out/godot-tools.js:144:22) %c(at t.onMessage.process.on.t.catch.t (/usr/lib/code/out/vs/workbench/services/extensions/node/extensionHostProcess.js:759:99))", source: file:///usr/lib/code/out/vs/workbench/workbench.main.js (240)
[1868:0828/122706.910784:INFO:CONSOLE(3267)] "[[object Object]]Cannot read property 'endsWith' of null", source: file:///usr/lib/code/out/vs/workbench/workbench.main.js (3267)
[1868:0828/122706.910815:INFO:CONSOLE(3267)] "TypeError: Cannot read property 'endsWith' of null
	at escape_command (/home/ascadian/.vscode-oss/extensions/geequlim.godot-tools-0.9.0/out/godot-tools.js:60:31)
	at run_godot (/home/ascadian/.vscode-oss/extensions/geequlim.godot-tools-0.9.0/out/godot-tools.js:70:34)
	at run_editor.Promise (/home/ascadian/.vscode-oss/extensions/geequlim.godot-tools-0.9.0/out/godot-tools.js:96:17)
	at new Promise (<anonymous>)
	at GodotTools.run_editor (/home/ascadian/.vscode-oss/extensions/geequlim.godot-tools-0.9.0/out/godot-tools.js:55:16)
	at open_workspace_with_editor.Promise (/home/ascadian/.vscode-oss/extensions/geequlim.godot-tools-0.9.0/out/godot-tools.js:45:22)
	at new Promise (<anonymous>)
	at GodotTools.open_workspace_with_editor (/home/ascadian/.vscode-oss/extensions/geequlim.godot-tools-0.9.0/out/godot-tools.js:38:16)
	at retry_connect_client.vscode.window.showErrorMessage.then.item (/home/ascadian/.vscode-oss/extensions/geequlim.godot-tools-0.9.0/out/godot-tools.js:144:22)", source: file:///usr/lib/code/out/vs/workbench/workbench.main.js (3267)

```
This issue seems similar:
https://github.com/microsoft/vscode-maven/issues/337
With this fix:
https://github.com/microsoft/vscode-maven/pull/381/files


#### Starting the project closes the editor
When I start the editor via the lsp dialog stating that godot is not running and I start the game from vscode afterward, the editor is closed.

### My configuration
I compiled godot from this commit:
https://github.com/ankitpriyarup/godot/commit/149180508993cacaa4220ded59088353eab3fa12
My VS Code is:
```
Version: 1.36.1
Commit: 2213894ea0415ee8c85c5eea0d0ff81ecc191529
Date: 2019-07-22T07:38:21.043Z
Electron: 4.2.8
Chrome: 69.0.3497.128
Node.js: 10.11.0
V8: 6.9.427.31-electron.0
OS: Linux x64 5.1.15-1-surface
```
Extension version:
https://github.com/ankitpriyarup/godot-tools/commit/46ba5a1ac2fe728c8bbb0cea9983e8c92c27895d

akien-mga commented on modules/gdscript/language_server/gdscript_language_protocol.h:
```diff
@@ -0,0 +1,93 @@
+#include "lsp.hpp"
+#include "modules/jsonrpc/jsonrpc.h"
+#include "modules/websocket/websocket_peer.h"
+#include "modules/websocket/websocket_server.h"
```
Note that this will fail building if either the jsonrpc or websocket modules are disabled. We don't have a good system to define module inter-dependencies, but this should be addressed eventually.

Especially since building Godot without websocket but with gdscript might be a valid use case, so it should be made possible to build gdscript without the LSP in this case.

akien-mga replied:
Solved by #31741.

akien-mga commented:
I'll merge this PR as is so that @ankitpriyarup can submit their additional (re)work.

We should then review if @ankitpriyarup's work solves the comments which have been raised in this PR or not, and if not, we should ensure that all of them are handled further on (and marked as resolved when they are).

akien-mga commented:
Thanks a lot!

Geequlim commented:
@Schroedi Thanks for your testing.
I tested your project with the master LSP server the code completion doesn't work with `s1.gd` and `s2.gd` which contain cyclic reference errors. But works for `s3.gd` which does not contains errors.

The bug about shell should be fixed in the coming days.

Thanks for your test. You can report related  issuses here [godotengine/godot-vscode-plugin](https://github.com/godotengine/godot-vscode-plugin)
